tests/ui/unnecessary_lazy_eval.rs

```diff
@@ -181,10 +183,34 @@ fn main() {
     // neither bind_instead_of_map nor unnecessary_lazy_eval applies here
     let _: Result<usize, usize> = res.and_then(|x| Err(x));
     let _: Result<usize, usize> = res.or_else(|err| Ok(err));
+
+    issue9422(3);
+    panicky_arithmetic_ops(3);
 }
 
 #[allow(unused)]
 fn issue9485() {
     // should not lint, is in proc macro
     with_span!(span Some(42).unwrap_or_else(|| 2););
 }
+
+fn issue9422(x: usize) -> Option<usize> {
+    (x >= 5).then(|| x - 5)
+    // (x >= 5).then_some(x - 5)  // clippy suggestion panics
+}
+
+// https://doc.rust-lang.org/stable/reference/expressions/operator-expr.html#overflow
+fn panicky_arithmetic_ops(x: usize) {
+    let _x = false.then(|| i32::MAX + 1); // don't lint
+    let _x = false.then(|| i32::MAX * 2); // don't lint
+    let _x = false.then(|| i32::MAX - 1); // lint
+    #[allow(clippy::identity_op)]
+    let _x = false.then(|| (1 + 2 * 3 - 2 / 3 + 9) << 2); // lint
+    let _x = false.then(|| 255u8 << 7); // lint
+    let _x = false.then(|| 255u8 << 8); // don't lint
+    let _x = false.then(|| 255u8 >> 8); // don't lint
+    let _x = false.then(|| 255u8 >> x); // don't lint
```

Comment:
This is the only line actually using the argument to this function (all other are using literals and/or constants); what about replacing this `x` with e.g. `-7` (negative), removing the parameter, and adding a new function taking `x: i32` (signed) and testing things like `(x != 0).then(|| i32::MAX / x)`, `(x != i32::MIN).then(|| -x)` etc.?

Reply:
It's been a while but, FWIW, those other lines that don't use the parameter are there to actually check that the updated const eval logic works the way it should with panicky arithmetic in mind (i.e., when it would or wouldn't panic due to overflow at runtime).

When clippy's const eval sees a path to a local like the `x` in the expression `255 >> x` we're passing to it, it immediately bails out of the evaluator, so I guessed there wouldn't be much of a need/wouldn't be interesting to test different kinds of variations of this and thought one case would be enough, as it just immediately drops out in any case and doesn't even get to the interesting stuff. So for anything involving a local variable such as a parameter, this isn't gonna change eager to lazy or lazy to eager and thus will not emit a lint. Unless I misunderstood your comment?

The test case `255 >> -7` sounds like itd be worth testing though. I haven't yet looked at the other comment but from a quick glance this case (negative RHS) is basically what you were commenting on, I think.

Reply:
Ah OK, I had ticked at the lone variable in the midst of consts (then suggested adding some "more realistic" uses as variations of the original issue) but that's only to test the "bail out of const eval" thing.

👍 for the added test